sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+        checkArgument(
+            streamExists(spec.getAWSClientsProvider().getKinesisClient(), spec.getStreamName()),
+            "Stream %s does not exist", spec.getStreamName());
+
+        // Init producer config
+        Properties props = spec.getProducerProperties();
+        if (props == null) {
+          props = new Properties();
+        }
+        KinesisProducerConfiguration config = KinesisProducerConfiguration.fromProperties(props);
+        // Fix to avoid the following message "WARNING: Exception during updateCredentials" during
+        // producer.destroy() call. More details can be found in this thread:
+        // https://github.com/awslabs/amazon-kinesis-producer/issues/10
+        config.setCredentialsRefreshDelay(100);
+
+        // Init Kinesis producer
+        producer = spec.getAWSClientsProvider().createKinesisProducer(config);
+        // Use custom partitioner if it exists
+        if (spec.getPartitioner() != null) {
+          partitioner = spec.getPartitioner();
+        }
       }
 
-      @Override
-      public AmazonKinesis getKinesisClient() {
-        AmazonKinesisClientBuilder clientBuilder =
-            AmazonKinesisClientBuilder.standard().withCredentials(getCredentialsProvider());
-        if (serviceEndpoint == null) {
-          clientBuilder.withRegion(region);
-        } else {
-          clientBuilder.withEndpointConfiguration(
-              new AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region.getName()));
+      /**
+       * It adds a record asynchronously which then should be delivered by Kinesis producer in
+       * background (Kinesis producer forks native processes to do this job).
+       *
+       * <p>The records can be batched and then they will be sent in one HTTP request. Amazon KPL
+       * supports two types of batching - aggregation and collection - and they can be configured by
+       * producer properties.
+       *
+       * <p>More details can be found here:
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html">KPL Key Concepts</a> and
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-config.html">Configuring the KPL</a>
+       */
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        ByteBuffer data = ByteBuffer.wrap(c.element());
+
+        String partitionKey = spec.getPartitionKey();
+        String explicitHashKey = null;
+
+        // Use custom partitioner
+        if (partitioner != null) {
+          partitionKey = partitioner.getPartitionKey(c.element());
+          explicitHashKey = partitioner.getExplicitHashKey(c.element());
         }
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
+      }
+
+      @FinishBundle
```

Comment:
Does anything need to be done in `StartBundle`?

Reply:
I don't think so since we create producer in `Setup`. Perhaps, in `StartBundle`, we can check  if it's not null (I'm not sure that it's possible) and create new instance if needed. What do you think?